### ISSUE TITLE
Support for embedded structured data (metadata) in the body

### ISSUE THREAD

**Issue by gsergiu** [w3c/web-annotation#309]
I think there is a common need of embedding structured data in the bodies. A very simple example is the geo-tagging, having the goal to relate a target resource to a map based presentation layer. (the google imagery [Johan Strauss Statue in Stadtpark - Viena](https://www.google.at/maps/place/Stadtpark/@48.202975,16.378018,3a,75y,90t/data=!3m8!1e2!3m6!1s-ryowUlRVgCo%2FVKG2wP9fj3I%2FAAAAAAAACN8%2FgW_65eo2KOsso42_pJT7r1SJrW_m8vO8Q!2e4!3e12!6s%2F%2Flh5.googleusercontent.com%2F-ryowUlRVgCo%2FVKG2wP9fj3I%2FAAAAAAAACN8%2FgW_65eo2KOsso42_pJT7r1SJrW_m8vO8Q%2Fs203-k-no%2F!7i2048!8i1238!4m5!3m4!1s0x476d0775d3324143:0xbf209e4513aec145!8m2!3d48.2047697!4d16.3805724!6m1!1e1))
Therefore I would suggest to add the EmbeddedDataset type to the bodies.
Example:

```
“body”: {
    "@context" : "<geo_context.jsonld>", 
    “type”:[Dataset, Geolocation],
    “latitude”: ”<x>”,
    “longitude”: “<y>”,
    "title":"Johan Strauss Statue in Stadtpark - Viena",
    “format”:”application/vnd.google-earth.kml+xml”
} 
```

**Comment by gsergiu** [comment]
PS: of course a `@context` element needs to be added to the body to teach clients how to interpret the  latitude, longitude and title.

**Comment by iherman** [comment]
The documents are now under feature freeze, I have therefore assigned the 'postpone' label and added to the V2 milestone.

**Comment by gsergiu** [comment]
:( ... maybe this is something to add to the "long living" document. Are there any news about the fromal process for V2? I think that the beaurocratic process of generating the version 1 of the standard should not stop the progress towards V2 (as soon as the main stakeholders are not blocked by finalization of V1 text).

**Comment by iherman** [comment]
The WG has not discussed yet on how it proposes to W3C to handle V2. All what I can say is a personal opinion for now: one way is that, after this WG closes, a community group is created (or the old one 'rechartered') in discussing, distilling, etc, what would be in a V2, and then a WG is formed to finalize it. Alternatively, this WG would be rechartered with the explicit goal of a V2. There are precedents for both approaches.

**Comment by gsergiu** [comment]
Just for the record, I think that the second option makes more sense in the curent situation. 
The number of recorded open issues for V2 could be a strong argument for it
Of course this implies that the members of the group want to continue this work.

**Comment by azaroth42** [comment]
I suggest close -- this is already possible via extensions. We're never going to enumerate all of the properties that might be wanted (like latitude and longitude in the example).

**Comment by iherman** [comment]
I am fine with closing

**Comment by gsergiu** [comment]
Yes, it is true. The geotagging is an extension (or two?).
However, I think that there is no guidance for now to:
1. Embedd metadata in the Body
2. Represent geolocations

I think that both these aspects are very important and they are very common use-cases (in the sense that thee are a lot of applications that have the goal to put objects on the map, but very few of them use annotation standards).

> We're never going to enumerate all of the properties that might be wanted (like latitude and longitude in the example).

Well ... even if the title is broad with the intention to illustrate the extension mechanism with a concrete example, the usecase is very concrete, and there are enough standards to represent geolocations:
like GML for example: [GML](https://en.wikipedia.org/wiki/Geography_Markup_Language).

So ... of-course, one cannot specify the support for any possible metadata and standards, but I think that it will be very important to illustrate the mechanism of extending the Body to be able to carry embedded metadata represented in an existing (official or defacto) standard format. (The mechanism should be the for other types of extensions)

**Comment by azaroth42** [comment]
Is there something else needed beyond what's in the extensions section of the vocabulary already?
  http://w3c.github.io/web-annotation/vocab/wd/#extensions

**Comment by jjett** [comment]
Isn't this use case solved by just supplying a named graph that contains all of the metadata as the body or by simply linking the body to a metadata description using a dc:description predicate as a part of the local implementation? I'm not certain I understand the concern here.

**Comment by gsergiu** [comment]
> Isn't this use case solved by just supplying a named graph that contains all of the metadata

@jjett I think that the named graph can be a good solution to solve the issue. However, there are a few concerns regarding the "embedded structured data":
1. standard JSON parsers should be able to understand the body as an object that carries structured data.  (currently `type:Dataset` and the missing of `id` or `source` is to be interpreted as structured data)
2. dc:description and dc:format would be recommended to describe the "payload", the metadata.
3. The "structured data" might be mapped to json using a jsonld context, but it could be simply embedded in the "value" using the original formating (e.g. xml).  
4. The structured data shouldn't be mixed with regular body properties. (e.g. the header(body props)-payload ("structured-data") kind of structure should be used, becasue the body and the "structured" data have different schemas, and might have different formats (e.g. json body, xml payload))
5. The schema needed to parse the "structured data" should be written (as URI) in the body object prefferably (i.e. json client should be able to send payload to a presentation layer that is able to render the data). 
   6.Which is the recommeded way to make this information avalailable? Should the body use 2 types, e.g.  "type":["Dataset","Geolocation"], so that any standard annotation parser is able to understand that this is embedded structured data, and it needs to sent it to a presentation layer that is able to render Geolocations?

PS: I realize now that the are 2 major approches:
1. map schema of the "structured data" to json,
2. or embedd using original format (e.g. text/xml) 

Main question: Should the standard support both approaches?

I would say yes!
For mapping schemas through jsonld context, the examples would be fairly similar to the 107: Extension Example 2. (if the mapping is made directly into the body and not in a named graph)

BR,
Sergiu

**Comment by gsergiu** [comment]
@azaroth42 

> Is there something else needed beyond what's in the extensions section of the vocabulary already?
> http://w3c.github.io/web-annotation/vocab/wd/#extensions
1. In this section I would like to see an example for adding extensions to the body (i.e. the existing ones extend the annotation and the target).
2. Could we have an example of using a named graph for extending the target or the body?
3. The specific resource looks like a minified representation of a named graph. Am I right? Is there any guidance about when to use named graphs (i.e. `@graph`) and when to use specific reources?

**Comment by iherman** [comment]
I have the impression that we are getting into a primer territory and not a specification territory (even if there is a fine line between the two). We may want to have a separate discussion whether there is enough interest (and time!) to produce a primer or something. I think that the specification itself should not be extended beyond what it has now.

**Comment by gsergiu** [comment]
@iherman yes, you are rigth, I'm analyzing the issue in a larger context. But there are still smaller issue closely related to the current version of the standard:
1. Could we add an example of extensions for the body?
2. Could this example use the "type":"Dataset"? I think that embeded Dataset is quite a common/important usecase which is probably supported by the current specifications, but there is no example of not making it explicit
3. There is only one mention of the named graphs in the stadard, is the following recommendation still standing, especially in the context of extensions? How should we interpret it, this is a bit unclear to me? 
   `Deprecate embedded graphs as an explicit part of the model, instead just include or reference a serialized graph.`

**Comment by iherman** [comment]
> On 21 Jun 2016, at 11:28, gsergiu notifications@github.com wrote:
> 
> @iherman https://github.com/iherman yes, you are rigth, I'm analyzing the issue in a larger context. But there are still smaller issue closely related to the current version of the standard:
> 
> Could we add an example of extensions for the body?
> Could this example use the "type":"Dataset"? I think that embeded Dataset is quite a common/important usecase which is probably supported by the current specifications, but there is no example of not making it explicit
> There is only one mention of the named graphs in the stadard, is the following recommendation still standing, especially in the context of extensions? Deprecate embedded graphs as an explicit part of the model, instead just include or reference a serialized graph.

@gsergiu, I am extremely conservative at this point, mindful of the Working Group's schedules, editorial jobs, priorities. So my preference is _not_ to add anything to the current specs.

**Comment by BigBlueHat** [comment]
@gsergiu I'm going to :+1: @iherman's thoughts on routing this toward Primer related content.

Additionally, you have the full power of JSON-LD (and RDF for that matter) at your disposal when extending your own Web Annotation documents.

A Primer document could certainly highlight that, and having some examples on hand (perhaps built by you? :wink:) would be super great.

At the spec-level, though, we should leave things alone to make sure it makes it out the door in the first place.

**Comment by gsergiu** [comment]
@iherman  @BigBlueHat 
I do understand the reluctance to add new things in the specifications at this stage, and I think that the some things reported here do not necessarily need to be addressed in the specifications of the first version of the standard. However my personal opinion is that there are some things that should be added/enhanced in the first version of the draft:
1.  The extension for the targets are likely to happen through selectors and a concrete example is provided. Anyway, I think that the most common case will be the extension of the body to represent domain specific annotations. I strongly believe that an example for extending the body should be included in the vocabulary (whatever usecase the editors think is the easiest to include)
2.  The “logical” types of the body are ExternalResource,  SpecificResource, BodyValue, TextualBody (embeded text) ... and I think the (embedded) Dataset/StructuredData type is missing here. As it is no time now to add the type in the Specifications, at least a non-normative Note should be added to notify about this possible type. 
3.  I think the sentence “Deprecate embedded graphs as an explicit part of the model, instead just include or reference a serialized graph.”  Must be improved. 
I’m not sure if I’m correct or not, but I assume that this is related to the SpecificResources and the design decision for the model/vocabulary. 
- However, the sentence is hard to be interpreted as so, especially because there is no reference to the “named graphs” in any of the specification documents.   
- Is this the official recommendation for the extensions as well? What is the difference between “embedded graph” and “serialized graph”? I assume that a graph needs to be serialized in order to be included, isn’t it so? This means that there is already a sentence, and there is no explanation of what an “embedded graph” and a “serialized graph” means in the context of WA.

**Comment by iherman** [comment]
Procedural comment: any _editorial_ change, that constitutes simply clarification, can be done for the Proposed Rec version of the specification and for that only. The CR version _is closed_. I do not want to interfere with the discussion regarding this (I must admit I do not have a strong feeling either way, here).

Administratively, if the consensus is to do this for the Proposed Rec, then the milestone should be changed to V1 PR.

The same comment goes to #308.

**Comment by azaroth42** [comment]
Re 1.  The extensions section explicitly says that 

> "Extensions may be made to any resource, including adding new objects as well as properties."
> and the three examples are clearly described as examples.  I'm sorry that your particular use case didn't get covered in the examples, but extensions can go anywhere. Contriubtions to a primer or recipe book would be appreciated, but there's no requirement to change the specification at this point.

Re 2.  There isn't a class "oa:ExternalResource". That would always be relative to an Annotation, which is not how the open world works. They're just Resources.  Datasets are interesting, but unlikely to be a significant use case in the wild.  They can be created already as bodies, no need to specify anything further.

Re 3. We lack experience, implementations and use cases for named graphs.  It also goes against the design principle of limiting the exposure to RDF concepts in order to implement. We discussed this already in the WG (a lot!) and there's no new information to warrant re-opening it for version 1.  I'll open a new issue tagged as version 2 to rediscuss.  If there were two independent implementations of the feature, that would go a long way towards demonstrating the utility and importance, however at this stage implementations of the features we do have is more valuable.

(Following discussion on telco of 2016-06-23)

**Comment by gsergiu** [comment]
Hi Rob, 
1. I didn't ment that the specifications don't allow extensions for the body. But the examples of extensions don't include an example for extendind the body, which I would say it is is the most important part of the annotation.  I really feel that this is a part where I feel the specifications to be incomplete. As I said I do realize that there is no time now to include this in V1, but it seems that there are more parts of the standard that really need extensions when implementing support for "domain specific" annotations. As suggested in other tickets, it would be good to have a "living" document (or primer) referenced in V1 and ... developed in V2.  
2. Yes .. there isn't a class oa:ExternalResource, as the consequence of preferring the "simplified" representation over the specific resources (it seems the we have to leave with that .. and it's drawbacks). Regarding the Dataset type, I would claim the contrary. If the type of any web api is Dataset (as there is no other matching type for apis), that I would claim the Dataset should be the most important type! However, this is just a remarc on indicated point where the document is  not "complete" in explanations/example, but there is no limitation in the text for this aspect. (if the editors consider that there is no reason to add a non-normative note in V1, ... we have to leave with the decision (.. as we have to leave with the brexit :( ) )
3. Well .. I don't find the sentence on named graphs in the text anymore (vocabulary or model). I just wanted to suggest its removal, as "named graph" subject is not used anywhere else in the text. (ok for me)
